File: backend/app/db/models/menu.py
import uuid

from sqlalchemy import (
    Boolean,
    CheckConstraint,
    Column,
    ForeignKey,
    Index,
    Integer,
    String,
    Table,
    Text,
)
from sqlalchemy.dialects.postgresql import UUID as PG_UUID
from sqlalchemy.orm import Mapped, mapped_column, relationship

from app.db.base import Base, SoftDeleteMixin, TimestampMixin, UUIDPrimaryKeyMixin

product_categories = Table(
    "product_categories",
    Base.metadata,
    Column(
        "product_id",
        PG_UUID(as_uuid=True),
        ForeignKey("products.id", ondelete="CASCADE"),
        primary_key=True,
    ),
    Column(
        "category_id",
        PG_UUID(as_uuid=True),
        ForeignKey("categories.id", ondelete="CASCADE"),
        primary_key=True,
    ),
)


class Category(UUIDPrimaryKeyMixin, TimestampMixin, SoftDeleteMixin, Base):
    __tablename__ = "categories"

    restaurant_id: Mapped[uuid.UUID] = mapped_column(
        PG_UUID(as_uuid=True),
        ForeignKey("restaurants.id", ondelete="CASCADE"),
        nullable=False,
    )
    name: Mapped[str] = mapped_column(Text, nullable=False)
    description: Mapped[str | None] = mapped_column(Text, nullable=True)
    image_path: Mapped[str | None] = mapped_column(Text, nullable=True)
    sort_index: Mapped[int] = mapped_column(Integer, nullable=False, server_default="0")

    products: Mapped[list["Product"]] = relationship(
        secondary=product_categories, back_populates="categories"
    )

    __table_args__ = (Index("ix_categories_listing", "restaurant_id", "is_active", "sort_index"),)


class Product(UUIDPrimaryKeyMixin, TimestampMixin, SoftDeleteMixin, Base):
    __tablename__ = "products"

    restaurant_id: Mapped[uuid.UUID] = mapped_column(
        PG_UUID(as_uuid=True),
        ForeignKey("restaurants.id", ondelete="CASCADE"),
        nullable=False,
    )
    name: Mapped[str] = mapped_column(Text, nullable=False)
    description: Mapped[str | None] = mapped_column(Text, nullable=True)
    price_cents: Mapped[int] = mapped_column(Integer, nullable=False)
    currency: Mapped[str] = mapped_column(String(3), nullable=False, server_default="USD")
    image_path: Mapped[str | None] = mapped_column(Text, nullable=True)
    approval_status: Mapped[str] = mapped_column(String, nullable=False, server_default="draft")
    is_published: Mapped[bool] = mapped_column(Boolean, nullable=False, server_default="false")

    categories: Mapped[list["Category"]] = relationship(
        secondary=product_categories, back_populates="products"
    )
    option_groups: Mapped[list["OptionGroup"]] = relationship(
        back_populates="product", cascade="all, delete-orphan"
    )

    __table_args__ = (
        CheckConstraint(
            "approval_status IN ('draft','pending_review','approved','rejected')",
            name="approval_status_allowed",
        ),
        Index("ix_products_publish", "restaurant_id", "is_active", "is_published"),
        Index("ix_products_review", "restaurant_id", "approval_status"),
    )


class OptionGroup(UUIDPrimaryKeyMixin, TimestampMixin, Base):
    __tablename__ = "option_groups"

    product_id: Mapped[uuid.UUID] = mapped_column(
        PG_UUID(as_uuid=True),
        ForeignKey("products.id", ondelete="CASCADE"),
        nullable=False,
    )
    title: Mapped[str] = mapped_column(Text, nullable=False)
    required: Mapped[bool] = mapped_column(Boolean, nullable=False, server_default="false")
    selection: Mapped[str] = mapped_column(String, nullable=False)
    min_selections: Mapped[int] = mapped_column(Integer, nullable=False, server_default="0")
    max_selections: Mapped[int | None] = mapped_column(Integer, nullable=True)
    sort_index: Mapped[int] = mapped_column(Integer, nullable=False, server_default="0")
    is_active: Mapped[bool] = mapped_column(Boolean, nullable=False, server_default="true")

    product: Mapped["Product"] = relationship(back_populates="option_groups")
    items: Mapped[list["OptionItem"]] = relationship(
        back_populates="option_group", cascade="all, delete-orphan"
    )

    __table_args__ = (
        CheckConstraint("selection IN ('single','multi')", name="selection_allowed"),
        Index("ix_option_groups_product", "product_id"),
    )


class OptionItem(UUIDPrimaryKeyMixin, TimestampMixin, Base):
    __tablename__ = "option_items"

    option_group_id: Mapped[uuid.UUID] = mapped_column(
        PG_UUID(as_uuid=True),
        ForeignKey("option_groups.id", ondelete="CASCADE"),
        nullable=False,
    )
    label: Mapped[str] = mapped_column(Text, nullable=False)
    price_delta_cents: Mapped[int] = mapped_column(Integer, nullable=False, server_default="0")
    sort_index: Mapped[int] = mapped_column(Integer, nullable=False, server_default="0")
    is_active: Mapped[bool] = mapped_column(Boolean, nullable=False, server_default="true")

    option_group: Mapped["OptionGroup"] = relationship(back_populates="items")

    __table_args__ = (Index("ix_option_items_group", "option_group_id"),)

File: frontend/src/lib/api/menu.ts
import { DEFAULT_CURRENCY } from '@/lib/currency';
import { apiRequest } from './client';
import type { Category, CategoryDisplayLayout, CursorPage, Product } from './types';

export function listCategories(
  token: string,
  restaurantId: string,
  limit: number,
  cursor?: string | null,
  options?: { includeInactive?: boolean },
) {
  const params = new URLSearchParams({ limit: String(limit) });
  if (cursor) params.set('cursor', cursor);
  if (options?.includeInactive) params.set('include_inactive', 'true');
  return apiRequest<CursorPage<Category>>(
    `/restaurants/${restaurantId}/categories?${params}`,
    { token },
  );
}

export function listProducts(
  token: string,
  restaurantId: string,
  limit: number,
  cursor?: string | null,
) {
  const params = new URLSearchParams({ limit: String(limit) });
  if (cursor) params.set('cursor', cursor);
  return apiRequest<CursorPage<Product>>(
    `/restaurants/${restaurantId}/products?${params}`,
    { token },
  );
}

export function createCategory(
  token: string,
  restaurantId: string,
  data: {
    name: string;
    description?: string | null;
    image_path?: string | null;
    sort_index?: number;
  },
) {
  return apiRequest<Category>(`/restaurants/${restaurantId}/categories`, {
    method: 'POST',
    token,
    body: {
      restaurant_id: restaurantId,
      name: data.name,
      description: data.description ?? null,
      image_path: data.image_path ?? null,
      sort_index: data.sort_index ?? 0,
    },
  });
}

export function updateCategory(
  token: string,
  restaurantId: string,
  categoryId: string,
  data: {
    name?: string;
    description?: string | null;
    image_path?: string | null;
    sort_index?: number;
    display_layout?: CategoryDisplayLayout | null;
    is_active?: boolean;
  },
) {
  return apiRequest<Category>(
    `/restaurants/${restaurantId}/categories/${categoryId}`,
    { method: 'PATCH', token, body: data },
  );
}

export function setCategoryProductOrder(
  token: string,
  restaurantId: string,
  categoryId: string,
  productIds: string[],
) {
  return apiRequest<void>(
    `/restaurants/${restaurantId}/categories/${categoryId}/products/order`,
    { method: 'PUT', token, body: { product_ids: productIds } },
  );
}

export function getProduct(token: string, restaurantId: string, productId: string) {
  return apiRequest<Product>(`/restaurants/${restaurantId}/products/${productId}`, { token });
}

export function createProduct(
  token: string,
  restaurantId: string,
  data: {
    name: string;
    description?: string | null;
    price_cents: number;
    currency?: string;
    image_path?: string | null;
    approval_status?: string;
    is_published?: boolean;
    category_ids: string[];
  },
) {
  return apiRequest<Product>(`/restaurants/${restaurantId}/products`, {
    method: 'POST',
    token,
    body: {
      restaurant_id: restaurantId,
      currency: DEFAULT_CURRENCY,
      approval_status: 'approved',
      is_published: true,
      ...data,
    },
  });
}

export function updateProduct(
  token: string,
  restaurantId: string,
  productId: string,
  data: {
    name?: string;
    description?: string | null;
    price_cents?: number;
    currency?: string;
    image_path?: string | null;
    approval_status?: string;
    is_published?: boolean;
    is_active?: boolean;
    category_ids?: string[];
  },
) {
  return apiRequest<Product>(
    `/restaurants/${restaurantId}/products/${productId}`,
    { method: 'PATCH', token, body: data },
  );
}

export type OptionGroupCreateInput = {
  title: string;
  required: boolean;
  selection: 'single' | 'multi';
  min_selections?: number;
  max_selections?: number | null;
  sort_index?: number;
  is_active?: boolean;
  items: { label: string; price_delta_cents: number; sort_index?: number }[];
};

export function createOptionGroup(
  token: string,
  restaurantId: string,
  productId: string,
  data: OptionGroupCreateInput,
) {
  return apiRequest<Product['option_groups'][number]>(
    `/restaurants/${restaurantId}/products/${productId}/option-groups`,
    {
      method: 'POST',
      token,
      body: data,
    },
  );
}

export type OptionGroupUpdateInput = {
  title?: string;
  required?: boolean;
  selection?: 'single' | 'multi';
  min_selections?: number;
  max_selections?: number | null;
  sort_index?: number;
  is_active?: boolean;
};

export function updateOptionGroup(
  token: string,
  restaurantId: string,
  productId: string,
  groupId: string,
  data: OptionGroupUpdateInput,
) {
  return apiRequest<Product['option_groups'][number]>(
    `/restaurants/${restaurantId}/products/${productId}/option-groups/${groupId}`,
    { method: 'PATCH', token, body: data },
  );
}

export function deleteOptionGroup(
  token: string,
  restaurantId: string,
  productId: string,
  groupId: string,
) {
  return apiRequest<void>(
    `/restaurants/${restaurantId}/products/${productId}/option-groups/${groupId}`,
    { method: 'DELETE', token },
  );
}

export function createOptionItem(
  token: string,
  restaurantId: string,
  productId: string,
  groupId: string,
  data: { label: string; price_delta_cents: number; sort_index?: number },
) {
  return apiRequest<Product['option_groups'][number]['items'][number]>(
    `/restaurants/${restaurantId}/products/${productId}/option-groups/${groupId}/items`,
    { method: 'POST', token, body: data },
  );
}

export function deleteOptionItem(
  token: string,
  restaurantId: string,
  productId: string,
  groupId: string,
  itemId: string,
) {
  return apiRequest<void>(
    `/restaurants/${restaurantId}/products/${productId}/option-groups/${groupId}/items/${itemId}`,
    { method: 'DELETE', token },
  );
}

export type OptionItemUpdateInput = {
  label?: string;
  price_delta_cents?: number;
  sort_index?: number;
  is_active?: boolean;
};

export function updateOptionItem(
  token: string,
  restaurantId: string,
  productId: string,
  groupId: string,
  itemId: string,
  data: OptionItemUpdateInput,
) {
  return apiRequest<Product['option_groups'][number]['items'][number]>(
    `/restaurants/${restaurantId}/products/${productId}/option-groups/${groupId}/items/${itemId}`,
    { method: 'PATCH', token, body: data },
  );
}
